### ISSUE TITLE
打撃音検知モデルのためのシンプル改善案

### ISSUE THREAD

**Issue by Kalorie560** [Kalorie560/SoundDitect#57]
@claude
以下の通り、モデルの改善を目指します。
目的:
現在の1D CNNの性能を、打撃音の特性に合わせてシンプルかつ効果的に向上させる。

改善方針:

カーネルサイズの最適化: 最初の層で大きなカーネルを使い「響き」を、後の層で小さなカーネルを使い「アタック」の特徴を捉える。
必須要素の追加: CNNの標準的な構成要素であるMaxPooling層を導入し、次元削減と位置のズレに対する頑健性を高める。
学習の安定化: 学習率を自動で調整し、不均衡なデータを考慮した損失関数を用いる。

例えば、Yamlファイル内の以下の部分について改善を検討してください。
```
model:
  architecture: "1d_cnn_attention"
  input_length: 44100
  num_classes: 2
  
  cnn_layers:
    # 【変更点①】カーネルサイズを階層的に設定（大 -> 小）
    # 最初の層で大きなカーネルを使い、打撃音の響きや低周波成分を捉える
    - filters: 64
      kernel_size: 255 # 約5.8msの範囲。音の響きを捉えるため広めに設定
      stride: 1
      padding: "same"
    # 【コード側での実装指示】この直後にMaxPooling1D(pool_size=4)を追加

    - filters: 128
      kernel_size: 63  # 中くらいの範囲で特徴を絞り込む
      stride: 1
      padding: "same"
    # 【コード側での実装指示】この直後にMaxPooling1D(pool_size=4)を追加

    - filters: 128
      kernel_size: 31  # 急峻なアタック部分など、細かい特徴を捉える
      stride: 1
      padding: "same"
    # 【コード側での実装指示】この直後にMaxPooling1D(pool_size=4)を追加
  
  attention:
    hidden_dim: 128
    num_heads: 4
  
  fully_connected:
    - units: 256
      dropout: 0.3
    - units: 128
      dropout: 0.3

training:
  batch_size: 4
  epochs: 50

  # 【変更点②】学習率スケジューラの導入
  lr_scheduler:
    name: "CosineAnnealingLR" # 学習率を自動調整し、安定した収束を目指す
    T_max: 50
    eta_min: 0.00001
  
  # 【変更点③】不均衡データ対策
  loss_function:
    name: "binary_crossentropy"
    # 打撃音（異常）データが少ない場合、その重みを上げて検知しやすくする
    # 例：正常データが異常の5倍なら、pos_weight: 5.0
    pos_weight: 5.0

  validation_split: 0.2
  early_stopping:
    patience: 10 # 以前より短くして、効率的に学習
    monitor: "val_loss"
```

YAMLのコメントにある通り、各cnn_layersの畳み込み処理の後にMaxPooling1D層を追加してください。
これが性能向上に大きく貢献します。

**Comment by Kalorie560** [comment]
@claude
padding : "same" でエラーが発生したため、適切な値を代入。
requirements.txtは以下の内容でないとエラーが出るので修正。
```
# SoundDitect Training Dependencies
# Model training and generation for JSON data

# Essential Dependencies
pyyaml==6.0.1

# macOS LibreSSL Compatibility
urllib3==1.26.18
requests==2.32.3
certifi==2024.2.2
pyOpenSSL==23.3.0

# AI/ML Dependencies
torch==2.6.0
torchaudio==2.6.0
numpy==2.0.0
scipy==1.14.0
librosa==0.10.1
scikit-learn==1.5.0

# Experiment Tracking
clearml==1.13.2

# Data Processing
pandas==2.2.2
soundfile==0.12.1

# Utilities
python-dotenv==1.0.0
click==8.1.7
tqdm==4.67.0

# Development Dependencies (optional)
pytest==7.4.3
black==23.11.0
flake8==6.1.0

# Visualization
matplotlib==3.9.0
```
